RPi sensor events

DanielDecker opened the pull request:
Added support for detecting press duration for the GPIO sensor.
Modified GPIO actuator to allow toggle commands, added status update for all configured connections to keep them up to date. Added filters to ignore own messages and ones which no effect (output status already set).
Updated the readme's.
Tested the changes in-depth for all possible connection types.
Fixes #88

rkoshak reviewed (commented):
Run the code through pylint against pep8 and fix any big findings.  That helps with long term maintainability.

If we can, I'd like to have the filter behavior apply to all but I'm not certain of the implications. If there is something that would break then we can leave it with the two separate publish functions.

DanielDecker commented on core/actuator.py:
```diff
@@ -78,3 +78,10 @@ def _publish(self, message, destination):
             conn.publish(message, destination)
+            
+    def _publish_actuator_state(self, message, destination):
+        """This method will pulisch the given message as a actuator update
```
fixed

DanielDecker commented on gpio/README.md:
```diff
@@ -81,7 +81,13 @@ Parameter | Required | Restrictions | Purpose
+`Destination` | X | | Location/openHAB string item to publish the pin state as OPEN/CLOSED (default).
+`Values` | | | Values to replace OPEN,CLOSED message as comma separeted list. Eg. `OFF,ON`
+`Short_Press-Dest` | | | Location/openHAB string/datetime item to publish an update after a short button press happend. Which are two chages of the logic level at the selected pin. Eg. with `PUD = DOWN` LOW > HIGH > LOW and the duration between the edges is between `Short_Press-Threshold` and `Long_Press-Threshold`. Works best with `EventDetection = BOTH`
+`Short_Press-Threshold` | | decimal number | Devines the lower bound of short button press event in seconds, if the duration of the button press was shorter no update will be send. Usful to ignor false detection of button press due to electrical interferences. (default is 0)
```
fixed

DanielDecker commented on gpio/README.md:
```diff
@@ -81,7 +81,13 @@ Parameter | Required | Restrictions | Purpose
+`Short_Press-Dest` | | | Location/openHAB string/datetime item to publish an update after a short button press happend. Which are two chages of the logic level at the selected pin. Eg. with `PUD = DOWN` LOW > HIGH > LOW and the duration between the edges is between `Short_Press-Threshold` and `Long_Press-Threshold`. Works best with `EventDetection = BOTH`
+`Short_Press-Threshold` | | decimal number | Devines the lower bound of short button press event in seconds, if the duration of the button press was shorter no update will be send. Usful to ignor false detection of button press due to electrical interferences. (default is 0)
+`Long_Press-Dest` | | | Location/openHAB string/datetime item to publish an update after a long button press happend, requires `Long_Press-Threshold`, `Short_Press-Dest`
+`Long_Press-Threshold` | | decimal number | Devines the lower bound of long button press event in seconds, if the duration of the button press was shorter a short button event will be triggered
```
fixed

DanielDecker commented on gpio/README.md:
```diff
@@ -104,21 +110,26 @@ Pin = 17
 ## `gpio.rpi_gpio.RpiGpioActuator`
 
 Commands a GPIO pin to go high, low, or if configured with a toggle it goes high for half a second and then goes to low.
+A recieved command will be send back on all configured connections to the configured `CommandSrc`, to keep them up to date.
```
fixed.
Thanks for spell checking. :-)

DanielDecker commented on gpio/README.md:
```diff
@@ -134,13 +145,15 @@ $ sudo pip3 install RPI.GPIO
 `Connection` | X | Comma separated list of Connections | Where the ON/OFF messages are published.
 `Level` | | DEBUG, INFO, WARNING, ERROR | When provided, sets the logging level for the sensor.
 `CommandSrc` | X | | Destination/openHAB switch item where commands are received, expects ON/OFF. If Toggle is set all messages trigger a toggle.
+`ToggleCommandSrc` | | | Destination/openHAB  string/datetime item where toggle commands are recieverd. This is intended to be used for direct connections to a sensor via the Short_Press-Dest/Long_Press-Dest parameter. Expects the string TOGGLE or the current time as java formated datetime value e.g. `2021-10-31T17:03:20.829945`. When one of those are recieved output of the actuator will get togg ...
```
First of all with this code there are two ways to make the actuator toggle the output. Sending the string "TOOGLE" which works fine if only one connection is involved. If for some reason a sensor and an actuator are connected with more then one connection the sensor would send TOOGLE on each connection and the actuator would toggle each time since it doesn't know the command was intended only once. Therefore the gpio sensor will send a datetime to Short_Press-Dest and the actuator will only toggle if the datetime changes.
ToggleCommandSrc is intended to receive toggle commands either via the string "TOGGLE" or via datetime

rkoshak replied:
Maybe it would be more intuitive and require less work on the devices part  to make add a debounce parameter instead of sending date times. Then you can code in a time period after receiving a TOGGLE where you ignore all subsequent TOGGLE commands. 

I can think of lots of cases where needing to publish an ISO8601 string would be a really big deal and even where that's not a really big deal, it requires that the time be perfectly synchronized across all these potential sources of the command. If commands can indeed come from more than one device they almost certainly will never have the same time down to the msec level and in fact could be many minutes or hours apart in cases where there is no real time clock. On an Arduino and other microcontrollers the only clock is the number of msecs since the device started up.

DanielDecker commented on core/actuator.py:
```diff
@@ -78,3 +78,10 @@ def _publish(self, message, destination):
         for conn in self.connections:
             conn.publish(message, destination)
+            
+    def _publish_actuator_state(self, message, destination):
```
Using this filter for the general 
`def publish(self, message, destination):`
function would break local connections and MQTT connections on the same device (local sensor connected with local actuator via MQTT).

So the only way to join `publish` with `publish_actuator_state` would be to add another parameter to publish. Since I didn't wanted to rewrite other parts of the project which rely on `publish` and I don't know if overloading of the function is possible or desired I made a separate function.

rkoshak replied:
Sounds reasonable. I'll keep that in mind next time I decide to to a refactor/rewrite. ;-)

DanielDecker commented:
I'll check the code with pylint as soon I got it working with eclipse :-/

DanielDecker commented:
Today I figured that publishing the current actuator state on a `refresh` request, the way it works sensors, would be useful. I'll add this in the actuator class, as an optional component, as soon I get your replay on my other comments.

rkoshak commented:
You can run pylint from the command line if needed in a pinch. 

I'm pretty happy with everything except the date time string. I don't think it will be reliable or even possible to work if the intent is to truly support cases where more than one device could command the toggle based on the same event. I think you can get to the same goal using a debounce parameter and simply ignore commands that come too close after the first one. It doesn't have to be long, maybe a quarter of a second would be enough.

DanielDecker commented:
I fixed the pylint problems which had been introduced by my code.
The rpi actuator works now as suggested: only one toggle command is accepted during the debounce period. Added config parameter ToggleDebounce with default value 0.05 (seconds). On my test setup this value worked well.
To avoid confusion of the toggle command with the toggle parameter I renamed the parameter to 'ButtonSimulation'. Old configs with `Toggle = true` are recognized anyway.

Known Bug:
When an sensor is connected to an actuator with param `Toggle = true` via local connection the button press duration will be detected as 0.5s or greater.

Any more suggestions?

rkoshak commented:
I think it looks good now. Thanks!
